### PR TITLE
vm: allow parsing scientific JSON numbers

### LINKED ISSUE #3069: T5 testnet block 2418704 processing issue
blockQueue: failed adding block into the blockchain        {"error": "onPersist failed: VM has failed: at instruction 0 (SYSCALL): insufficient funds", "blockHeight": 2418703, "nextIndex": 2418704}

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #3069.

The first commit fixes state difference at block 2336911@T5. After that the new state difference at block 2342940@T5 occurs:
```
state differs at 2342940, block 254985ed16cb865742971a93b16aa6888aa20ccc7082d56f9e59d2e53ca151ea
transaction 4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938:
--- http://localhost:20332
+++ http://seed1t5.neo.org:20332
@@ -1,2 +1,2 @@
-(*result.ApplicationLog)(0xc00030f7c0)({
+(*result.ApplicationLog)(0xc000039840)({
Container: (util.Uint256) (len=32 cap=32) 38996258c81e1a4a0ebeaa04fa1a54fe959187820f187ea7e25ab79bb2339e4c,
@@ -9,3 +9,3 @@
Stack: ([]stackitem.Item) (len=1 cap=1) {
-    (*stackitem.Array)(0xc0002625a0)(Array)
+    (*stackitem.Array)(0xc0003e2840)(Array)
},

different state found
exit status 1
```

Something tells me that it's still the same JSON-related issue, but from some other side of view, see the new numbers at the 105 instruction (`"[9.05e+28,1.871e+21,3.0366e+32,1e+30]"`):
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go query tx 4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938 -r http://localhost:20332 -v
Hash:			4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938
OnChain:		true
BlockHash:		254985ed16cb865742971a93b16aa6888aa20ccc7082d56f9e59d2e53ca151ea
Success:		true
Signer:			NMQUeV1hPhNCM4k8x8dXFA25Y8oiDER8yZ (CalledByEntry)
SystemFee:		0.03349697 GAS
NetworkFee:		0.00132852 GAS
Script:			DEBeQh5b5nPV+E55v8+REMpGUqYjEioobzZQ6QSbLQTmCuLHMBJEdCEGkml/wSeg3HT+YKInSL5D0l6M72Kas6XBEcAMIQI3Q0to88Qvxv82bpGWEPmmjbsyesk4dtdqegUcoK8nZhHADCVbOS4wNWUrMjgsMS44NzFlKzIxLDMuMDM2NmUrMzIsMWUrMzBdE8AfDAt1cGRhdGVQcmljZQwUb6vCut1+dKyebIdJvOFmVK9KBbtBYn1bUg==
INDEX    OPCODE       PARAMETER
0        PUSHDATA1    5e421e5be673d5f84e79bfcf9110ca4652a623122a286f3650e9049b2d04e60ae2c730124474210692697fc127a0dc74fe60a22748be43d25e8cef629ab3a5c1    <<
66       PUSH1
67       PACK
68       PUSHDATA1    0237434b68f3c42fc6ff366e919610f9a68dbb327ac93876d76a7a051ca0af2766
103      PUSH1
104      PACK
105      PUSHDATA1    5b392e3035652b32382c312e383731652b32312c332e30333636652b33322c31652b33305d ("[9.05e+28,1.871e+21,3.0366e+32,1e+30]")
144      PUSH3
145      PACK
146      PUSH15
147      PUSHDATA1    7570646174655072696365 ("updatePrice")
160      PUSHDATA1    6fabc2badd7e74ac9e6c8749bce16654af4a05bb
182      SYSCALL      System.Contract.Call (627d5b52)
```

Go node stack:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ curl -d '{ "jsonrpc": "2.0", "id": 1, "method": "getapplicationlog", "params": ["4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938"] }' localhost:20332 | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   642  100   502  100   140   245k  70000 --:--:-- --:--:-- --:--:--  313k
{
   "result" : {
      "txid" : "0x4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938",
      "executions" : [
         {
            "gasconsumed" : "3316532",
            "trigger" : "Application",
            "notifications" : [],
            "exception" : null,
            "vmstate" : "HALT",
            "stack" : [
               {
                  "type" : "Array",
                  "value" : [
                     {
                        "value" : "90499999999999999999932891136",
                        "type" : "Integer"
                     },
                     {
                        "type" : "Integer",
                        "value" : "1871000000000000000000"
                     },
                     {
                        "type" : "Integer",
                        "value" : "303659999999999999994193741086720"
                     },
                     {
                        "value" : "1000000000000000000024696061952",
                        "type" : "Integer"
                     }
                  ]
               }
            ]
         }
      ]
   },
   "jsonrpc" : "2.0",
   "id" : 1
}
```

C# node stack:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ curl -d '{ "jsonrpc": "2.0", "id": 1, "method": "getapplicationlog", "params": ["4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938"] }' seed1t5.neo.org:20332 | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   641    0   501  100   140    532    148 --:--:-- --:--:-- --:--:--   680
{
   "id" : 1,
   "result" : {
      "executions" : [
         {
            "notifications" : [],
            "gasconsumed" : "3316532",
            "exception" : null,
            "vmstate" : "HALT",
            "stack" : [
               {
                  "type" : "Array",
                  "value" : [
                     {
                        "value" : "90499999999999993918259200000",
                        "type" : "Integer"
                     },
                     {
                        "value" : "1871000000000000000000",
                        "type" : "Integer"
                     },
                     {
                        "value" : "303660000000000004445016810323968",
                        "type" : "Integer"
                     },
                     {
                        "value" : "1000000000000000019884624838656",
                        "type" : "Integer"
                     }
                  ]
               }
            ],
            "trigger" : "Application"
         }
      ],
      "txid" : "0x4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938"
   },
   "jsonrpc" : "2.0"
}
```

Thus, **the PR is not finished**, I'll keep investigating.

### CODE REVIEW

File: pkg/vm/stackitem/json.go [hunk 1]
```diff
@@ -213,18 +213,35 @@ func (d *decoder) decode() (Item, error) {
...
-			for _, r := range ts[dot+1:] {
-				if r != '0' {
-					return nil, fmt.Errorf("%w (real value for int)", ErrInvalidValue)
+		var (
+			num *big.Int
+			ok  bool
+		)
+		isScientific := strings.Contains(ts, "e+") || strings.Contains(ts, "E+")
+		if isScientific {
+			// As a special case numbers like 2.8e+22 are allowed (SetString rejects them).
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/stackitem/json.go:222]: It's a float value, it shouldn't be parsed (and it won't). We have the `Exact` check for this.

**Comment by AnnaShaleva** [comment]
It turns out that increasing the digint parsing precision from the default up to maximum allowed (`big.MaxPrec`, as in the current version of this PR) helps and the resulting NeoGo node stack for the second statediff transaction is the following:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ curl -d '{ "jsonrpc": "2.0", "id": 1, "method": "getapplicationlog", "params": ["4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938"] }' http://localhost:20332 | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   642  100   502  100   140   163k  46666 --:--:-- --:--:-- --:--:--  208k
{
   "result" : {
      "txid" : "0x4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938",
      "executions" : [
         {
            "vmstate" : "HALT",
            "exception" : null,
            "notifications" : [],
            "trigger" : "Application",
            "gasconsumed" : "3316532",
            "stack" : [
               {
                  "value" : [
                     {
                        "type" : "Integer",
                        "value" : "90500000000000000000000000000"
                     },
                     {
                        "type" : "Integer",
                        "value" : "1871000000000000000000"
                     },
                     {
                        "type" : "Integer",
                        "value" : "303660000000000000000000000000000"
                     },
                     {
                        "value" : "1000000000000000000000000000000",
                        "type" : "Integer"
                     }
                  ],
                  "type" : "Array"
               }
            ]
         }
      ]
   },
   "id" : 1,
   "jsonrpc" : "2.0"
}
```

These values are exactly the same values that user provides as an input to `updatePrice` method: `[9.05e+28,1.871e+21,3.0366e+32,1e+30]`. And these values are stored in the contract storage which affects the state. However, now I see three problems here:
1. The C# node values differ from the expected, which means inaccuracy in big numbers parsing on the C# node side.
2. We can't provide the same solid integers parsing for values that are bigger than `big.MaxPrec` allows.
3. C# node allows to unmarshal floats from scientific notation (like `1.2345e+3`), and it must be no-op. However, I've checked the current testnet, there are no such situations when someone provides floating point scientific number to StdLib's deserializeJSON.

I've described the proposed solution in the https://github.com/neo-project/neo/issues/2879.

File: pkg/vm/stackitem/json.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/stackitem/json.go:222]: Bingo!

**Comment by AnnaShaleva** [comment]
Technically, this PR makes everything that possible to makes things better. But we won't go further than 2342940@T5 without https://github.com/neo-project/neo/issues/2879 resolution.

File: pkg/vm/stackitem/json.go [hunk 2]
```diff
@@ -213,18 +213,35 @@ func (d *decoder) decode() (Item, error) {
...
+			// That's the way how C# code works.
+			f, _, err := big.ParseFloat(ts, 10, big.MaxPrec, big.ToNearestEven)
```
Comment (roman-khimov) [inline pkg/vm/stackitem/json.go:224]: We have a 256-bit VM, it should be sufficient instead of `big.MaxPrec`.

**Comment by AnnaShaleva** [comment]
States match up to the current T5 height:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ go run scripts/compare-states/compare-states.go http://localhost:20332 http://seed1t5.neo.org:20332
at 0: 9c1361fbbe2a0ce31910f992cbdfb8afee6f178e5d1aabdd1c6f2af90440e239 vs 9c1361fbbe2a0ce31910f992cbdfb8afee6f178e5d1aabdd1c6f2af90440e239
at 2441424: f00416df2782e5bc3c7d7b92cabd81e17a2b75089a0652460bc164dde4c60a19 vs f00416df2782e5bc3c7d7b92cabd81e17a2b75089a0652460bc164dde4c60a19
```
Mainnet check is in progress.

File: pkg/vm/stackitem/json.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/vm/stackitem/json.go]: Temporary set to `53` to match the ref. implementation. Testnet is OK with that, mainnet is being checked.

**Comment by AnnaShaleva** [comment]
Mainnet states match up to the current height:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ go run ./scripts/compare-states/compare-states.go http://localhost:10332 http://seed1.neo.org:10332
at 0: bb2bd2dfee742024b966cdf757c6235044cb9e0ce43497ad0eb27c49b8bdf4fe vs bb2bd2dfee742024b966cdf757c6235044cb9e0ce43497ad0eb27c49b8bdf4fe
at 3849893: 35519c007dfaefea7c483f93eca989ed9b5f2be05707ef7cdf8779987713fec9 vs 35519c007dfaefea7c483f93eca989ed9b5f2be05707ef7cdf8779987713fec9
```

**Comment by AnnaShaleva** [comment]
Ready to be released.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**